synapse/rest/client/receipts.py

```diff
@@ -109,6 +109,46 @@ async def on_POST(
 
         return 200, {}
 
+    async def _is_valid_thread_id(self, event_id: str, thread_id: str) -> bool:
+        """
+        The thread ID provided must relate (in a vague sense) to the event ID.
+
+        We check this to ensure clients aren't sending bogus receipts.
+
+        A thread ID is considered valid if:
+
+        1. The event has a thread relation which matches the thread ID.
+        2. The event has children events which form a thread relation (i.e. the
+           event is a thread root).
+        2. The event is related to an event (recursively) which satisfies 1 or 2.
+
+        Given the following DAG:
+
+            A <---[m.thread]-- B <--[m.annotation]-- C
+            ^
+            |--[m.reference]-- D <--[m.annotation]-- E
```

Comment:
What is m.reference? I can't see its meaning described in https://spec.matrix.org/v1.4/client-server-api/

Reply:
See matrix-org/matrix-spec#1206. (Note that annotations aren't even accepted: matrix-org/matrix-spec-proposals#2677, but they're in a weird spot and generally considered supported...I think having it here is fine.)